File: code/ch04/4.1.1.CountAndSay.cpp
class Solution {
public:
    string countAndSay(int n) {
        string res = "1";
        
        for (int i = 1; i < n; i++) {
            string tmp = "";
            char current_char = res[0];
            int  char_count   = 0;
            
            for (int j = 0; j < res.size(); j++) {
                if (res[j] != current_char) {
                    tmp += itoa(char_count) + current_char;
                    current_char = res[j];
                    char_count = 1;
                } else {
                    char_count += 1;
                }
            }
            
            tmp += itoa(char_count) + current_char;
            res = tmp;
        }
        
        return res;
    }
};

File: code/ch04/4.1.1.CountAndSay.py
class Solution:
    def countAndSay(self, n: int) -> str:
        ans = "1"
        for i in range(1, n):
            tmp = ""
            current_char, char_count = ans[0], 0
            for j in range(len(ans)):
                if ans[j] != current_char:
                    tmp += str(char_count) + current_char
                    current_char, char_count = ans[j], 1
                else:
                    char_count += 1
            tmp += str(char_count) + current_char
            ans = tmp
        return ans

